nearswap/src/twap.rs

```diff
@@ -0,0 +1,189 @@
+            price0cumulative: U128(price0),
+            price1cumulative: U128(price1),
+        });
+        return 0;
+    }
+
+    /// @notice Writes an oracle observation to the array
+    /// @dev Index represents the most recently written element.
+    /// max_length and index must be tracked externally.
+    /// @param observation The stored TWAP array
+    /// @param index The location of the most recently updated observation
+    /// @param block_timestamp The timestamp of the new observation
+    /// @param max_length The length of TWAP array
+    /// @return updated index
+    pub fn write(
+        observation: &mut Vec<Twap>,
+        index: usize,
+        block_timestamp: u64,
+        price0: u128,
+        price1: u128,
+        max_length: usize
+    ) -> usize {
+        let last: &Twap = &observation[index];
+
+        let updated_index: usize = (index + 1) % max_length;
+        if index + 1 >= max_length {
+            observation[updated_index] = Twap::transform(last, block_timestamp, price0, price1);
+        } else {
+            observation.push(Twap::transform(last, block_timestamp, price0, price1));
+        }
+
+        return updated_index;
+    }
+
+    // Pivoted point binary search
+    // Similar to rotated array from a certain pivot point
+    // @param timestamp given timestamp
+    // @returns index of first timestamp that is greater than or equal to given timestamp
+    pub fn binary_search(
+        observation: &Vec<Twap>,
+        last_updated_index: usize,
+        max_length: usize,
+        block_timestamp: u64,
+        pivoted: bool
+    ) -> usize {
+        let mut start: usize = 0;
+        let mut end: usize = last_updated_index + 1;
+
+        let mut mid: usize;
+
+        while start < end {
+            mid = (start + end) / 2;
+            if u64::try_from(block_timestamp).unwrap() <= u64::try_from(observation[mid].block_timestamp).unwrap() {
+                end = mid;
+            } else {
+                start = mid + 1;
+            }
+        }
+
+        if(pivoted && start == 0) {
+            let res = start;
+            start = last_updated_index + 1;
+            end = max_length;
+
+            while start < end {
+                mid = (start + end) / 2;
+                if u64::try_from(block_timestamp).unwrap() <= u64::try_from(observation[mid].block_timestamp).unwrap() {
+                    end = mid;
+                } else {
+                    start = mid + 1;
+                }
+            }
+            if start == max_length - 1 
+                && u64::try_from(observation[start].block_timestamp).unwrap() < u64::try_from(block_timestamp).unwrap() {
+                start = res;
+            }
+
+            return start;
+        }
+
+        return start;
+    }
+
+    // function which will calculate mean using str "1min, 5min, 1h, 12h"
+    pub fn calculate_mean(
+        observation: &Vec<Twap>, time: &str, last_index: usize,
```

Comment:
when we will have a new Twap structure (and this structure will become Observation), then we should put that method in `Twap`, and use `self.observations`, instead of having it as an argument.

Reply:
ok